File: src/bin/jupiter_probe.rs
//! Verify the Jupiter Lend (Fluid) decoders against live mainnet and enumerate
//! every vault: collateral/debt pair, liquidation threshold, sizes, and a
//! first-pass liquidatable signal. Read-only.
//!
//! Detection honesty: precise per-price liquidatable detection needs Fluid's
//! tick↔price math (not reversed here). This reports the CONFIDENT on-chain
//! liquidation-activity flags (absorbed debt / branch_liquidated) and leaves the
//! authoritative check to the executor's liquidate simulation.
//!
//! Usage: HELIUS_RPC=<url> cargo run --release --bin jupiter_probe

use arb_engine::jupiter::{self, Vault, VaultConfig, VaultState};
use solana_pubkey::Pubkey;
use std::collections::HashMap;
use std::time::Duration;

fn rpc(endpoint: &str, body: serde_json::Value) -> Option<serde_json::Value> {
    for attempt in 0..4 {
        if let Ok(r) = ureq::post(endpoint).send_json(body.clone()) {
            if let Ok(v) = r.into_json::<serde_json::Value>() { return Some(v); }
        }
        std::thread::sleep(Duration::from_millis(400 << attempt));
    }
    None
}
fn b64(d: &serde_json::Value) -> Option<Vec<u8>> {
    use base64::Engine;
    base64::engine::general_purpose::STANDARD.decode(d.get(0)?.as_str()?).ok()
}
/// getProgramAccounts filtered by an 8-byte discriminator at offset 0.
fn gpa_by_disc(endpoint: &str, disc: &[u8; 8]) -> Vec<(Pubkey, Vec<u8>)> {
    let disc58 = bs58::encode(disc).into_string();
    let v = rpc(endpoint, serde_json::json!({"jsonrpc":"2.0","id":1,"method":"getProgramAccounts",
        "params":[jupiter::VAULTS_PROGRAM, {"encoding":"base64",
            "filters":[{"memcmp":{"offset":0,"bytes":disc58}}]}]}));
    let mut out = Vec::new();
    for e in v.as_ref().and_then(|v| v["result"].as_array()).into_iter().flatten() {
        if let (Some(pk), Some(data)) = (
            e["pubkey"].as_str().and_then(|s| s.parse::<Pubkey>().ok()),
            b64(&e["account"]["data"]),
        ) { out.push((pk, data)); }
    }
    out
}

fn main() {
    let _ = dotenvy::dotenv();
    let _ = rustls::crypto::ring::default_provider().install_default();
    let endpoint = std::env::var("HELIUS_RPC").or_else(|_| std::env::var("RPC_HTTP")).expect("HELIUS_RPC");

    // Decode all VaultConfig + VaultState, join by vault_id.
    let mut configs: HashMap<u16, (Pubkey, VaultConfig)> = HashMap::new();
    for (pk, d) in gpa_by_disc(&endpoint, &jupiter::VAULT_CONFIG_DISC) {
        if let Some(c) = VaultConfig::decode(&d) { configs.insert(c.vault_id, (pk, c)); }
    }
    let mut states: HashMap<u16, (Pubkey, VaultState)> = HashMap::new();
    for (pk, d) in gpa_by_disc(&endpoint, &jupiter::VAULT_STATE_DISC) {
        if let Some(s) = VaultState::decode(&d) { states.insert(s.vault_id, (pk, s)); }
    }
    println!("live: {} VaultConfig, {} VaultState decoded", configs.len(), states.len());

    let label = |m: &Pubkey| -> String {
        match m.to_string().as_str() {
            jupiter::USDC_MINT => "USDC".into(), jupiter::USDT_MINT => "USDT".into(),
            jupiter::WSOL_MINT => "wSOL".into(), s => s[..6].to_string(),
        }
    };

    let mut vaults: Vec<Vault> = Vec::new();
    for (vid, (cpk, c)) in &configs {
        if let Some((spk, s)) = states.get(vid) {
            vaults.push(Vault { config_pubkey: *cpk, state_pubkey: *spk, config: c.clone(), state: s.clone() });
        }
    }
    vaults.sort_by_key(|v| v.config.vault_id);

    let (mut n_usdc, mut n_usdt, mut n_sol, mut n_maybe) = (0, 0, 0, 0);
    println!("\n{:>3} {:>7} {:>7} {:>5} {:>5} {:>16} {:>16} {:>6} liq?", "vid", "collat", "debt", "CF%", "LT%", "tot_supply", "tot_borrow", "absorb");
    for v in &vaults {
        let c = &v.config; let s = &v.state;
        match c.debt_label() { "USDC" => n_usdc += 1, "USDT" => n_usdt += 1, "wSOL" => n_sol += 1, _ => {} }
        let maybe = v.maybe_liquidatable();
        if maybe { n_maybe += 1; }
        println!("{:>3} {:>7} {:>7} {:>5.1} {:>5.1} {:>16} {:>16} {:>6} {}",
            c.vault_id, label(&c.supply_token), label(&c.borrow_token),
            c.collateral_factor as f64 / 10.0, c.liquidation_threshold as f64 / 10.0,
            s.total_supply, s.total_borrow, s.absorbed_debt_amount,
            if maybe { "★MAYBE" } else { "" });
    }

    let in_scope = n_usdc + n_usdt + n_sol;
    println!("\n═══ summary ═══");
    println!("vaults: {}  | debt in-scope (USDC/USDT/SOL): {in_scope}  (USDC {n_usdc}, USDT {n_usdt}, wSOL {n_sol})", vaults.len());
    println!("VERIFIED: all {} vaults decode (pairs, thresholds, sizes) against live accounts.", vaults.len());
    println!("first-pass 'maybe liquidatable' (absorbed-debt > 0): {n_maybe}");
    println!("NOTE: precise per-price liquidatable detection needs Fluid tick↔price math (not");
    println!("      implemented); the executor's liquidate simulation is the ground-truth gate.");
}

File: src/bin/jupiterProbe.ts
// Port of src/bin/jupiter_probe.rs
//
// Verify the Jupiter Lend (Fluid) decoders against live mainnet and enumerate
// every vault: collateral/debt pair, liquidation threshold, sizes, and a
// first-pass liquidatable signal. Read-only.
//
// Detection honesty: precise per-price liquidatable detection needs Fluid's
// tick↔price math (not reversed here). This reports the CONFIDENT on-chain
// liquidation-activity flags (absorbed debt / branch_liquidated) and leaves the
// authoritative check to the executor's liquidate simulation.
//
// Usage: HELIUS_RPC=<url> tsx src/bin/jupiterProbe.ts

import 'dotenv/config';
import bs58 from 'bs58';
import { PublicKey } from '@solana/web3.js';
import { USDC_MINT, USDT_MINT, VAULT_CONFIG_DISC, VAULT_STATE_DISC, VAULTS_PROGRAM, Vault, VaultConfig, VaultState, WSOL_MINT } from '../lib/jupiter.js';

async function rpc(endpoint: string, body: unknown): Promise<any | undefined> {
  for (let attempt = 0; attempt < 4; attempt++) {
    try {
      const res = await fetch(endpoint, {
        method: 'POST',
        headers: { 'content-type': 'application/json' },
        body: JSON.stringify(body),
      });
      return await res.json();
    } catch {
      // fall through to retry
    }
    await new Promise((r) => setTimeout(r, 400 << attempt));
  }
  return undefined;
}

function b64(d: any): Buffer | undefined {
  const s = d?.[0];
  if (typeof s !== 'string') return undefined;
  return Buffer.from(s, 'base64');
}

/** getProgramAccounts filtered by an 8-byte discriminator at offset 0. */
async function gpaByDisc(endpoint: string, disc: Buffer): Promise<[PublicKey, Buffer][]> {
  const disc58 = bs58.encode(disc);
  const v = await rpc(endpoint, {
    jsonrpc: '2.0',
    id: 1,
    method: 'getProgramAccounts',
    params: [VAULTS_PROGRAM, { encoding: 'base64', filters: [{ memcmp: { offset: 0, bytes: disc58 } }] }],
  });
  const out: [PublicKey, Buffer][] = [];
  const arr: any[] = Array.isArray(v?.result) ? v.result : [];
  for (const e of arr) {
    const pkStr = e?.pubkey;
    const data = b64(e?.account?.data);
    if (typeof pkStr === 'string' && data !== undefined) {
      try {
        out.push([new PublicKey(pkStr), data]);
      } catch {
        // skip unparseable pubkey
      }
    }
  }
  return out;
}

async function main(): Promise<void> {
  const endpoint = process.env.HELIUS_RPC ?? process.env.RPC_HTTP;
  if (!endpoint) throw new Error('HELIUS_RPC');

  // Decode all VaultConfig + VaultState, join by vault_id.
  const configs = new Map<number, [PublicKey, VaultConfig]>();
  for (const [pk, d] of await gpaByDisc(endpoint, VAULT_CONFIG_DISC)) {
    const c = VaultConfig.decode(d);
    if (c !== undefined) configs.set(c.vaultId, [pk, c]);
  }
  const states = new Map<number, [PublicKey, VaultState]>();
  for (const [pk, d] of await gpaByDisc(endpoint, VAULT_STATE_DISC)) {
    const s = VaultState.decode(d);
    if (s !== undefined) states.set(s.vaultId, [pk, s]);
  }
  console.log(`live: ${configs.size} VaultConfig, ${states.size} VaultState decoded`);

  const label = (m: PublicKey): string => {
    const s = m.toBase58();
    if (s === USDC_MINT) return 'USDC';
    if (s === USDT_MINT) return 'USDT';
    if (s === WSOL_MINT) return 'wSOL';
    return s.slice(0, 6);
  };

  const vaults: Vault[] = [];
  for (const [vid, [cpk, c]] of configs) {
    const st = states.get(vid);
    if (st !== undefined) {
      const [spk, s] = st;
      const v = new Vault();
      v.configPubkey = cpk;
      v.statePubkey = spk;
      v.config = c;
      v.state = s;
      vaults.push(v);
    }
  }
  vaults.sort((a, b) => a.config.vaultId - b.config.vaultId);

  let nUsdc = 0;
  let nUsdt = 0;
  let nSol = 0;
  let nMaybe = 0;
  console.log(
    `\n${'vid'.padStart(3)} ${'collat'.padStart(7)} ${'debt'.padStart(7)} ${'CF%'.padStart(5)} ${'LT%'.padStart(5)} ${'tot_supply'.padStart(16)} ${'tot_borrow'.padStart(16)} ${'absorb'.padStart(6)} liq?`,
  );
  for (const v of vaults) {
    const c = v.config;
    const s = v.state;
    const dl = c.debtLabel();
    if (dl === 'USDC') nUsdc += 1;
    else if (dl === 'USDT') nUsdt += 1;
    else if (dl === 'wSOL') nSol += 1;
    const maybe = v.maybeLiquidatable();
    if (maybe) nMaybe += 1;
    console.log(
      `${String(c.vaultId).padStart(3)} ${label(c.supplyToken).padStart(7)} ${label(c.borrowToken).padStart(7)} ${(c.collateralFactor / 10.0).toFixed(1).padStart(5)} ${(c.liquidationThreshold / 10.0).toFixed(1).padStart(5)} ${s.totalSupply.toString().padStart(16)} ${s.totalBorrow.toString().padStart(16)} ${s.absorbedDebtAmount.toString().padStart(6)} ${maybe ? '★MAYBE' : ''}`,
    );
  }

  const inScope = nUsdc + nUsdt + nSol;
  console.log('\n═══ summary ═══');
  console.log(`vaults: ${vaults.length}  | debt in-scope (USDC/USDT/SOL): ${inScope}  (USDC ${nUsdc}, USDT ${nUsdt}, wSOL ${nSol})`);
  console.log(`VERIFIED: all ${vaults.length} vaults decode (pairs, thresholds, sizes) against live accounts.`);
  console.log(`first-pass 'maybe liquidatable' (absorbed-debt > 0): ${nMaybe}`);
  console.log("NOTE: precise per-price liquidatable detection needs Fluid tick↔price math (not");
  console.log("      implemented); the executor's liquidate simulation is the ground-truth gate.");
}

main().catch((e) => {
  console.error(e);
  process.exit(1);
});
